lib/CodeGen/MemoryAllocator.cpp

```diff
@@ -138,49 +154,626 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+    auto buffSizeIter = buffInfoArray[buffIdIter].size;
+    // Choose buffer with maximum size.
+    if (buffSizeIter > buffSizeMax) {
+      buffSizeMax = buffSizeIter;
+      buffIdMax = buffIdIter;
+    }
+    // If size is the same choose buffer with maximum live interval.
+    if (buffSizeIter == buffSizeMax) {
+      auto currTime = buffInfoArray[buffIdMax].timeStop -
+                      buffInfoArray[buffIdMax].timeStart;
+      auto iterTime = buffInfoArray[buffIdIter].timeStop -
+                      buffInfoArray[buffIdIter].timeStart;
+      if (iterTime > currTime) {
+        buffIdMax = buffIdIter;
+      }
+    }
+  }
+  return buffIdMax;
+}
+
+/// Memory allocation strategy based on the following logic:
+/// 1. Find maximum live size.
+/// 2. Find buffer with maximum live interval.
+/// 3. If multiple buffers with same live interval, find maximum size.
+static uint64_t
+MaxLiveSizeMaxBuffTime(size_t buffIdx,
+                       const std::vector<BuffInfo> &buffInfoArray,
+                       const std::vector<LiveInfo> &liveInfoArray) {
+  // Find maximum total live allocation.
+  const auto &liveBuffIdList = getMaxLiveSizeBuffIdList(liveInfoArray);
+  // Find buffer with maximum live interval within the maximum allocation.
+  assert(liveBuffIdList.size() && "Live buffer ID list is empty!");
+  uint64_t buffIdMax = 0;
+  uint64_t buffTimeMax = 0;
+  for (auto buffIdIter : liveBuffIdList) {
+    auto buffTimeIter = buffInfoArray[buffIdIter].timeStop -
+                        buffInfoArray[buffIdIter].timeStart;
+    // Choose buffer with maximum live interval.
+    if (buffTimeIter > buffTimeMax) {
+      buffTimeMax = buffTimeIter;
+      buffIdMax = buffIdIter;
+    }
+    // If live interval is the same choose buffer with maximum size.
+    if (buffTimeIter == buffTimeMax) {
+      auto currSize = buffInfoArray[buffIdMax].size;
+      auto iterSize = buffInfoArray[buffIdIter].size;
+      if (iterSize > currSize) {
+        buffIdMax = buffIdIter;
+      }
+    }
+  }
+  return buffIdMax;
+}
+
+/// Memory allocation strategy which allocates the segments in the same order as
+/// requested.
+static uint64_t SameOrder(size_t buffIdx,
+                          const std::vector<BuffInfo> &buffInfoArray,
+                          const std::vector<LiveInfo> &liveInfoArray) {
+  return buffIdx;
+}
+
+/// Array of available memory allocation strategies listed in the decreasing
+/// order of the likelihood of getting the best allocation efficiency. All these
+/// strategies are used one after the other in the exact order as listed here.
+/// In case one such strategy provides the maximum efficiency (best theoretical
+/// result) then the following ones (if any) are not used anymore.
+static std::vector<MemAllocStrategy> memAllocStrategies = {
+    MaxLiveSizeMaxBuffSize, MaxLiveSizeMaxBuffTime, SameOrder};
+
+/// Utility function to allocate all the segments at once using the given
```

Comment:
Please describe all parameters here and in other functions.

Reply:
Done.